File: packages/python/voyager_ogm/transaction.py
"""Voyager OGM Two-Layer Rollback Transaction & Context Manager.

Provides robust transaction and savepoint context managers that automatically
rollback in-memory entity mutations and AST allocations if an exception occurs.
"""

from __future__ import annotations

import itertools
from typing import TYPE_CHECKING

from voyager_ogm._voyager_rs import NativeTransaction

if TYPE_CHECKING:
    from types import TracebackType

_tx_counter = itertools.count(1)


class SavepointContext:
    """Context manager for a nested transaction savepoint.

    Automatically releases the savepoint on successful block exit,
    or rolls back in-memory entities to the savepoint if an exception occurs.

    Attributes:
        name: Name of the savepoint.
        tx: Parent Transaction instance.

    Example:
        >>> with tx.savepoint("sp1"):
        ...     # changes here are isolated to sp1
    """

    def __init__(self, tx: Transaction, name: str) -> None:
        """Initializes a SavepointContext.

        Args:
            tx: Parent transaction.
            name: Savepoint name.
        """
        self.tx = tx
        self.name = name

    def __enter__(self) -> SavepointContext:
        """Creates the named savepoint."""
        self.tx.create_savepoint(self.name)
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> bool:
        """Releases the savepoint on success or rolls back on exception."""
        if exc_type is not None:
            self.tx.rollback_to_savepoint(self.name)
            return False
        self.tx.release_savepoint(self.name)
        return False


class Transaction:
    """Two-Layer Transaction manager with in-memory savepoints and dirty rollback.

    Can be used explicitly or as a Python context manager (`with Transaction() as tx:`).

    Attributes:
        id: Unique transaction ID.
        state: Current state string ('ACTIVE', 'COMMITTED', 'ROLLED_BACK').
        is_active: Boolean indicating if transaction is open.

    Example:
        >>> with Transaction() as tx:
        ...     # perform mutations
        ...     pass # automatically committed on exit
    """

    def __init__(self, tx_id: int | None = None) -> None:
        """Initializes a new Transaction.

        Args:
            tx_id: Optional explicit transaction ID integer.
        """
        resolved_id = tx_id if tx_id is not None else next(_tx_counter)
        self._native = NativeTransaction(resolved_id)

    @property
    def id(self) -> int:
        """Returns the unique transaction identifier."""
        return self._native.id

    @property
    def state(self) -> str:
        """Returns the transaction lifecycle state ('ACTIVE', 'COMMITTED', 'ROLLED_BACK')."""
        return self._native.state

    @property
    def is_active(self) -> bool:
        """Checks if the transaction is actively accepting operations."""
        return self._native.is_active

    def savepoint(self, name: str) -> SavepointContext:
        """Creates a nested savepoint context manager.

        Args:
            name: Unique savepoint name.

        Returns:
            SavepointContext for use with `with tx.savepoint("name"):`.

        Raises:
            ValueError: If the transaction is not active.
        """
        if not self.is_active:
            raise ValueError("Cannot create savepoint on inactive transaction")
        return SavepointContext(self, name)

    def create_savepoint(self, name: str) -> None:
        """Explicitly creates a savepoint in the active transaction.

        Args:
            name: Unique savepoint name string.

        Raises:
            ValueError: If the transaction is not active.
        """
        self._native.savepoint(name)

    def rollback_to_savepoint(self, name: str) -> None:
        """Rolls back the active transaction state to a named savepoint.

        Args:
            name: Savepoint name string.

        Raises:
            ValueError: If the savepoint does not exist or transaction is inactive.
        """
        self._native.rollback_to_savepoint(name)

    def release_savepoint(self, name: str) -> None:
        """Releases a named savepoint from the transaction.

        Args:
            name: Savepoint name string.

        Raises:
            ValueError: If the savepoint does not exist or transaction is inactive.
        """
        self._native.release_savepoint(name)

    def commit(self) -> None:
        """Commits the transaction and clears pending mutations.

        Raises:
            ValueError: If the transaction is not active.
        """
        self._native.commit()

    def rollback(self) -> None:
        """Rolls back the transaction and reverts all in-memory mutations.

        Raises:
            ValueError: If the transaction is not active.
        """
        self._native.rollback()

    def __enter__(self) -> Transaction:
        """Enters the transaction context."""
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> bool:
        """Commits on clean exit, or rolls back if an exception occurred."""
        if exc_type is not None:
            if self.is_active:
                self.rollback()
            return False
        if self.is_active:
            self.commit()
        return False

File: crates/voyager-core/src/transaction.rs
//! Two-Layer Rollback Transaction and In-Memory Unit-of-Work.
//!
//! Provides two-phase transaction management ensuring in-memory AST arena handles,
//! entity mutations, and dirty property snapshots automatically revert upon database
//! transaction failures, aborts, or savepoint rollbacks.

use crate::ast::{LiteralValue, QueryAstArena};
use crate::error::{Error, Result};
use std::collections::BTreeMap;

/// State of an in-flight or completed transaction.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum TransactionState {
    /// Active transaction accepting mutations and savepoints.
    Active,
    /// Successfully committed transaction.
    Committed,
    /// Rolled back transaction with reverted in-memory state.
    RolledBack,
}

/// Recorded entity mutation for in-memory Unit-of-Work tracking and undo operations.
#[derive(Debug, Clone, PartialEq)]
pub enum EntityMutation {
    /// Newly created graph node entity.
    InsertNode {
        /// Generated client-side temporary identifier.
        temp_id: u64,
        /// Assigned graph node labels.
        labels: Vec<String>,
        /// Initial entity property map.
        properties: BTreeMap<String, LiteralValue>,
    },
    /// Modified existing graph node entity.
    UpdateNode {
        /// Unique node identifier.
        node_id: u64,
        /// Original property values prior to mutation (for undo rollback).
        old_properties: BTreeMap<String, LiteralValue>,
        /// Updated property values.
        new_properties: BTreeMap<String, LiteralValue>,
    },
    /// Deleted graph node entity.
    DeleteNode {
        /// Identifier of the deleted node.
        node_id: u64,
        /// Previous labels of the node (for restoration).
        old_labels: Vec<String>,
        /// Previous properties of the node (for restoration).
        old_properties: BTreeMap<String, LiteralValue>,
    },
    /// Newly created graph relationship edge.
    InsertEdge {
        /// Generated client-side temporary edge identifier.
        temp_id: u64,
        /// Source node identifier.
        source_id: u64,
        /// Target node identifier.
        target_id: u64,
        /// Relationship edge type name (e.g. "FOLLOWS").
        edge_type: String,
        /// Relationship property map.
        properties: BTreeMap<String, LiteralValue>,
    },
    /// Deleted graph relationship edge.
    DeleteEdge {
        /// Edge identifier.
        edge_id: u64,
        /// Source node identifier.
        source_id: u64,
        /// Target node identifier.
        target_id: u64,
        /// Relationship type name.
        edge_type: String,
        /// Previous properties prior to deletion.
        old_properties: BTreeMap<String, LiteralValue>,
    },
}

/// Snapshot of the memory arena and mutation log at a specific point in time.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CheckpointState {
    /// Length of the AST arena at checkpoint creation.
    pub arena_len: usize,
    /// Total count of recorded mutations in the Unit-of-Work at checkpoint creation.
    pub mutation_count: usize,
}

/// Named savepoint within an active transaction.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Savepoint {
    /// User-defined savepoint name.
    pub name: String,
    /// Captured checkpoint state.
    pub checkpoint: CheckpointState,
}

/// In-memory Unit of Work tracking entity lifecycle, dirty states, and pending mutations.
#[derive(Debug, Default, Clone)]
pub struct UnitOfWork {
    mutations: Vec<EntityMutation>,
    next_temp_id: u64,
}

impl UnitOfWork {
    /// Creates a new, empty Unit of Work.
    pub fn new() -> Self {
        Self {
            mutations: Vec::new(),
            next_temp_id: 1,
        }
    }

    /// Registers a newly created node in the unit of work.
    pub fn register_new_node(
        &mut self,
        labels: Vec<String>,
        properties: BTreeMap<String, LiteralValue>,
    ) -> u64 {
        let temp_id = self.next_temp_id;
        self.next_temp_id += 1;
        self.mutations.push(EntityMutation::InsertNode {
            temp_id,
            labels,
            properties,
        });
        temp_id
    }

    /// Registers an update to an existing node with original and updated properties.
    pub fn register_dirty_node(
        &mut self,
        node_id: u64,
        old_properties: BTreeMap<String, LiteralValue>,
        new_properties: BTreeMap<String, LiteralValue>,
    ) {
        self.mutations.push(EntityMutation::UpdateNode {
            node_id,
            old_properties,
            new_properties,
        });
    }

    /// Registers a node deletion with its prior labels and properties.
    pub fn register_deleted_node(
        &mut self,
        node_id: u64,
        old_labels: Vec<String>,
        old_properties: BTreeMap<String, LiteralValue>,
    ) {
        self.mutations.push(EntityMutation::DeleteNode {
            node_id,
            old_labels,
            old_properties,
        });
    }

    /// Registers a newly created relationship edge.
    pub fn register_new_edge(
        &mut self,
        source_id: u64,
        target_id: u64,
        edge_type: impl Into<String>,
        properties: BTreeMap<String, LiteralValue>,
    ) -> u64 {
        let temp_id = self.next_temp_id;
        self.next_temp_id += 1;
        self.mutations.push(EntityMutation::InsertEdge {
            temp_id,
            source_id,
            target_id,
            edge_type: edge_type.into(),
            properties,
        });
        temp_id
    }

    /// Registers a deleted relationship edge.
    pub fn register_deleted_edge(
        &mut self,
        edge_id: u64,
        source_id: u64,
        target_id: u64,
        edge_type: impl Into<String>,
        old_properties: BTreeMap<String, LiteralValue>,
    ) {
        self.mutations.push(EntityMutation::DeleteEdge {
            edge_id,
            source_id,
            target_id,
            edge_type: edge_type.into(),
            old_properties,
        });
    }

    /// Returns the sequence of all pending entity mutations.
    pub fn pending_mutations(&self) -> &[EntityMutation] {
        &self.mutations
    }

    /// Returns the total number of pending entity mutations.
    pub fn len(&self) -> usize {
        self.mutations.len()
    }

    /// Checks if there are no pending entity mutations.
    pub fn is_empty(&self) -> bool {
        self.mutations.is_empty()
    }

    /// Captures a snapshot checkpoint of the current unit of work and arena state.
    pub fn checkpoint(&self, arena: &QueryAstArena) -> CheckpointState {
        CheckpointState {
            arena_len: arena.len(),
            mutation_count: self.mutations.len(),
        }
    }

    /// Rolls back pending mutations and the AST arena to a prior checkpoint.
    pub fn rollback_to(&mut self, arena: &mut QueryAstArena, checkpoint: &CheckpointState) {
        arena.rollback_to(checkpoint.arena_len);
        if checkpoint.mutation_count < self.mutations.len() {
            self.mutations.truncate(checkpoint.mutation_count);
        }
    }

    /// Clears all pending mutations upon successful commit or transaction completion.
    pub fn clear(&mut self) {
        self.mutations.clear();
    }
}

/// Two-Layer Transaction Manager with in-memory savepoints and dirty rollback.
#[derive(Debug)]
pub struct Transaction {
    id: u64,
    state: TransactionState,
    initial_checkpoint: CheckpointState,
    savepoints: Vec<Savepoint>,
}

impl Transaction {
    /// Starts a new transaction capturing the initial unit of work and arena checkpoint.
    pub fn new(id: u64, uow: &UnitOfWork, arena: &QueryAstArena) -> Self {
        let initial_checkpoint = uow.checkpoint(arena);
        Self {
            id,
            state: TransactionState::Active,
            initial_checkpoint,
            savepoints: Vec::new(),
        }
    }

    /// Returns the unique transaction identifier.
    pub fn id(&self) -> u64 {
        self.id
    }

    /// Returns the current lifecycle state of the transaction.
    pub fn state(&self) -> TransactionState {
        self.state
    }

    /// Checks if the transaction is actively accepting operations.
    pub fn is_active(&self) -> bool {
        self.state == TransactionState::Active
    }

    /// Creates a named savepoint within the active transaction.
    ///
    /// # Errors
    ///
    /// Returns [`Error::TransactionError`] if the transaction is no longer active.
    pub fn savepoint(
        &mut self,
        name: impl Into<String>,
        uow: &UnitOfWork,
        arena: &QueryAstArena,
    ) -> Result<()> {
        if !self.is_active() {
            return Err(Error::TransactionError(
                "Cannot create savepoint on inactive transaction".into(),
            ));
        }
        let sp_name = name.into();
        let checkpoint = uow.checkpoint(arena);
        self.savepoints.push(Savepoint {
            name: sp_name,
            checkpoint,
        });
        Ok(())
    }

    /// Rolls back in-memory entities and AST arena to a named savepoint.
    ///
    /// # Errors
    ///
    /// Returns [`Error::TransactionError`] if savepoint does not exist or transaction is inactive.
    pub fn rollback_to_savepoint(
        &mut self,
        name: &str,
        uow: &mut UnitOfWork,
        arena: &mut QueryAstArena,
    ) -> Result<()> {
        if !self.is_active() {
            return Err(Error::TransactionError(
                "Cannot rollback savepoint on inactive transaction".into(),
            ));
        }

        let pos = self
            .savepoints
            .iter()
            .rposition(|sp| sp.name == name)
            .ok_or_else(|| {
                Error::TransactionError(format!("Savepoint '{name}' not found in transaction"))
            })?;

        let sp = self.savepoints[pos].clone();
        uow.rollback_to(arena, &sp.checkpoint);

        // Truncate savepoints created after this savepoint
        self.savepoints.truncate(pos + 1);
        Ok(())
    }

    /// Releases a named savepoint from the transaction.
    ///
    /// # Errors
    ///
    /// Returns [`Error::TransactionError`] if savepoint does not exist or transaction is inactive.
    pub fn release_savepoint(&mut self, name: &str) -> Result<()> {
        if !self.is_active() {
            return Err(Error::TransactionError(
                "Cannot release savepoint on inactive transaction".into(),
            ));
        }

        let pos = self
            .savepoints
            .iter()
            .rposition(|sp| sp.name == name)
            .ok_or_else(|| {
                Error::TransactionError(format!("Savepoint '{name}' not found in transaction"))
            })?;

        self.savepoints.remove(pos);
        Ok(())
    }

    /// Commits the transaction and marks its state as [`TransactionState::Committed`].
    ///
    /// # Errors
    ///
    /// Returns [`Error::TransactionError`] if transaction is already completed.
    pub fn commit(&mut self, uow: &mut UnitOfWork) -> Result<()> {
        if !self.is_active() {
            return Err(Error::TransactionError(
                "Cannot commit inactive transaction".into(),
            ));
        }
        self.state = TransactionState::Committed;
        uow.clear();
        self.savepoints.clear();
        Ok(())
    }

    /// Rolls back the entire transaction to its initial state, discarding all pending mutations.
    ///
    /// # Errors
    ///
    /// Returns [`Error::TransactionError`] if transaction is already completed.
    pub fn rollback(&mut self, uow: &mut UnitOfWork, arena: &mut QueryAstArena) -> Result<()> {
        if !self.is_active() {
            return Err(Error::TransactionError(
                "Cannot rollback inactive transaction".into(),
            ));
        }
        self.state = TransactionState::RolledBack;
        uow.rollback_to(arena, &self.initial_checkpoint);
        self.savepoints.clear();
        Ok(())
    }
}
